sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -0,0 +1,277 @@
+}
+
+uint32_t shutdown_redact(rom_error_t reason, shutdown_error_redact_t severity) {
+  uint32_t redacted = (uint32_t)reason;
+  if (reason == kErrorOk) {
+    return 0;
+  }
+  switch (severity) {
+    case kShutdownErrorRedactNone:
+      break;
+    case kShutdownErrorRedactError:
+      redacted &= ~kRomErrorMaskError;
+      break;
+    case kShutdownErrorRedactModule:
+      redacted &= ~(kRomErrorMaskError | kRomErrorMaskModule);
+      break;
+    case kShutdownErrorRedactAll:
+      redacted = kErrorUnknown;
+      break;
+    default:
+      redacted = kErrorUnknown;
+  }
+  return redacted;
+}
+
+SHUTDOWN_FUNC(shutdown_software_escalate(void)) {
+  // FIXME: Use a software alert when available.  For now, signal
+  // a fatal_intg_error from SRAM.
+  enum { kBase = TOP_EARLGREY_SRAM_CTRL_MAIN_BASE_ADDR };
+  abs_mmio_write32(kBase + SRAM_CTRL_ALERT_TEST_REG_OFFSET, 1);
+}
+
+SHUTDOWN_FUNC(shutdown_keymgr_kill(void)) {
+  enum {
+    kBase = TOP_EARLGREY_KEYMGR_BASE_ADDR,
+  };
+  uint32_t reg = bitfield_bit32_write(0, KEYMGR_CONTROL_START_BIT, true);
+  reg = bitfield_field32_write(reg, KEYMGR_CONTROL_DEST_SEL_FIELD,
+                               KEYMGR_CONTROL_DEST_SEL_VALUE_NONE);
+  reg = bitfield_field32_write(reg, KEYMGR_CONTROL_OPERATION_FIELD,
+                               KEYMGR_CONTROL_OPERATION_VALUE_DISABLE);
+  abs_mmio_write32(kBase + KEYMGR_CONTROL_REG_OFFSET, reg);
+  abs_mmio_write32(kBase + KEYMGR_SIDELOAD_CLEAR_REG_OFFSET, 1);
+}
+
+SHUTDOWN_FUNC(shutdown_sram_kill(void)) {
+  enum {
+    kBase = TOP_EARLGREY_SRAM_CTRL_MAIN_BASE_ADDR
+  };
+  // Make sure SRAM exec is disabled and locked.
+  abs_mmio_write32(kBase + SRAM_CTRL_EXEC_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + SRAM_CTRL_EXEC_REGWEN_REG_OFFSET, 0);
+
+  // Request a new scrambling key (making SRAM worthless), then lock.
+  uint32_t reg = bitfield_bit32_write(0, SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT, true);
+  abs_mmio_write32(kBase + SRAM_CTRL_CTRL_REG_OFFSET, reg);
+  abs_mmio_write32(kBase + SRAM_CTRL_CTRL_REGWEN_REG_OFFSET, 0);
+}
+
+SHUTDOWN_FUNC(shutdown_flash_kill(void)) {
+  enum {
+    kBase = TOP_EARLGREY_FLASH_CTRL_CORE_BASE_ADDR
+  };
+  // Disable access via the default region register.
+  // Configure region 0 as enabled, but with no access bits enabled.
+  // Configure the base and size to cover the entire flash region.
+  //
+  // FIXME: this didn't work for me.  Need to speak to tjaychen when he returns
+  // from vacation.
+  // 
+  //uint32_t reg = bitfield_bit32_write(0, FLASH_CTRL_MP_REGION_CFG_0_EN_0_BIT, true);
+  //reg = bitfield_field32_write(reg, FLASH_CTRL_MP_REGION_CFG_0_BASE_0_FIELD, 0);
+  //reg = bitfield_field32_write(reg, FLASH_CTRL_MP_REGION_CFG_0_SIZE_0_FIELD, 0x3ff);
+  //abs_mmio_write32(kBase + FLASH_CTRL_MP_REGION_CFG_0_REG_OFFSET, reg);
+}
+
+SHUTDOWN_FUNC(shutdown_epmp_kill(void)) {
+  // TODO: implement me.
+}
+
+SHUTDOWN_FUNC(shutdown_hang(void)) {
+#ifdef MOCK_ABS_MMIO
+#else
+  while (true) {
+    asm("1: wfi\n"
```

Comment:
Nit. Consider writing this like so to make it a little clearer:
```
asm volatile(
  "1:"
  "wfi; wfi; wfi; wfi; j 1b;"
  "wfi; wfi; j 1b;"
  "wfi; j 1b;"
);
```
This is personal taste but I find the common idiom of filling inline assembly with `\n`s very hard to read. (Also, this needs to be volatile to tell the compiler not to elide it.)